In [2]:
import pandas as pd
import nltk
import os

In [4]:
os.chdir(r'C:\Users\Patrick Stoyer\Dropbox\Documents\Sophomore Year\Semester_2\Ling_1340_Data_Science_for_Linguists\Project')

In [8]:
os.listdir()


['bbc_articles',
 'bbc_info.csv',
 'bbc_scrape.py',
 'kommersant_articles',
 'kommersant_info.csv',
 'kommersant_scrape.py',
 'radio_svoboda_articles',
 'radio_svoboda_info.csv',
 'radio_svoboda_scrape.py',
 'reuters_articles',
 'reuters_info.csv',
 'reuters_scrape.py',
 'tass_articles',
 'tass_info.csv',
 'tass_scrape.py']

FILES
----
The folder I moved to in the above code includes all the articles I've scraped, the code used to scrape them (note that each site used slightly different formatting, so I made separate scripts for each one), as well as a csv file that lists all the urls, titles, and dates (for tass, I still need to scrape most of the dates...the way I had to scrape it used selenium rather than bs4 because it had one of those infinite loading things and bc date is "hidden" within each article's source; this takes longer). Let's look at the csv files first

In [10]:
bbc_info = pd.read_csv('bbc_info.csv')
kommersant_info = pd.read_csv('kommersant_info.csv')
radio_svoboda_info = pd.read_csv('radio_svoboda_info.csv')
reuters_info = pd.read_csv('reuters_info.csv')
tass_info = pd.read_csv('tass_info.csv')

In [13]:
from IPython.core.interactiveshell import InteractiveShell
InteractiveShell.ast_node_interactivity = "all"


In [12]:
bbc_info.head()
kommersant_info.head()
radio_svoboda_info.head()
reuters_info.head()
tass_info.head()

Unnamed: 0                                              URL  \
0           0  https://www.bbc.co.uk/russian/features-47253360   
1           1      https://www.bbc.co.uk/russian/news-47260541   
2           2  https://www.bbc.co.uk/russian/features-47353363   
3           3  https://www.bbc.co.uk/russian/features-47135158   
4           4  https://www.bbc.co.uk/russian/features-47124214   

                                               Title             Date  
0  Россия никуда не денется зачем американцы учат...  21 февраля 2019  
1  Британский министр обороны Россия - угроза, но...  15 февраля 2019  
2    СМИ России что изменило убийство Бориса Немцова  25 февраля 2019  
3  Какими ракетами вооружатся Россия и США после ...   7 февраля 2019  
4  Видео\r\nТысяча сайтов в день Россия бьет реко...   5 февраля 2019

Unnamed: 0                                    URL  \
0           0  https://www.kommersant.ru/doc/3718636   
1           1  https://www.kommersant.ru/doc/3718635   
2           2  https://www.kommersant.ru/doc/3718611   
3           3  https://www.kommersant.ru/doc/3718600   
4           4  https://www.kommersant.ru/doc/3718243   

                                               Title        Date  
0  Фигурант по делу «Армии воли народа» Валерий П...  2018-08-18  
1  Володин не исключил встречу с Mail.ru по декри...  2018-08-18  
2  В Иркутской области сожгли автомобиль мэра гор...  2018-08-18  
3  Возбуждено уголовное дело по факту кражи нефти...  2018-08-18  
4  Рок и кантри объединяют людей, open space — от...  2018-08-18

Unnamed: 0                                      URL  \
0           0  https://www.svoboda.org/a/29785482.html   
1           1  https://www.svoboda.org/a/29785216.html   
2           2  https://www.svoboda.org/a/29785122.html   
3           3  https://www.svoboda.org/a/29785109.html   
4           4  https://www.svoboda.org/a/29785013.html   

                                               Title             Date  
0  После послания Путина социальные выплаты повыс...  22 Февраль 2019  
1  Сотрудника антикоррупционной полиции задержали...  22 Февраль 2019  
2  Суд арестовал москвичку, оставившую сына на но...  22 Февраль 2019  
3  Начата прокурорская проверка после срыва похор...  22 Февраль 2019  
4  Леонид Парфёнов перезапустит программу Намедни...  22 Февраль 2019

Unnamed: 0                                                URL  \
0           0  https://ru.reuters.com/article/topNews/idRUKCN...   
1           1  https://ru.reuters.com/article/topNews/idRUKCN...   
2           2  https://ru.reuters.com/article/topNews/idRUKCN...   
3           3  https://ru.reuters.com/article/topNews/idRUKCN...   
4           4  https://ru.reuters.com/article/topNews/idRUKCN...   

                                               Title        Date  
0  Европейские политики призвали к новым санкциям...  2018-27-11  
1  Судьба сделки о Brexit под вопросом на фоне кр...  2018-27-11  
2  Кремль опасается роста напряженности на юго-во...  2018-27-11  
3  Путин понизил планку приемлемой для РФ цены на...  2018-28-11  
4  Кремль встреча Путина и Трампа нужна обеим сто...  2018-28-11

Unnamed: 0                                   URL  \
0           0    https://tass.ru/obschestvo/6152766   
1           1    https://tass.ru/obschestvo/6152627   
2           2    https://tass.ru/obschestvo/6152545   
3           3  https://tass.ru/armiya-i-opk/6151531   
4           4      https://tass.ru/v-strane/6151365   

                                               Title         Date  
0  Марш памяти Бориса Немцова начался в центре Мо...  24 фев 2019  
1  Турпоток в заповедники Оренбургской области за...  24 фев 2019  
2  Первый фотоальбом лошадей Пржевальского из Оре...  24 фев 2019  
3  Северный флот представил историческую и соврем...  23 фев 2019  
4  Ограничение движения по зимнику из НАО в други...  23 фев 2019

In [19]:
bbc_info.shape
kommersant_info.shape
radio_svoboda_info.shape
reuters_info.shape
tass_info.shape

(350, 4)

(5290, 4)

(2929, 4)

(295, 4)

(3060, 4)

SIZE
----
kommersant is by far the largest corpus, followed by tass and radiosvoboda, and finally by bbc and reuters... reuters only had the most recent two months of articles (and only very few per day), bbc required me to search a keyword (i choose россия ==Russia) and limited the output to 350 articles... tass had many articles per day so the 3060 only goes back to like December (i could have gotten more, but once again, i was using selenium to scroll down the page and the fact that it returned about the same number of articles as radio svoboda was just chance...scrolling down to a specific date would be harder... for kommersant, i went back to the same date that radio svoboda's articles started from...radio svoboda only let me go back 100 pages, and had 29 articles per day...i think i messed up and scraped the first page twice so i'll have to fix that eventually.


In [23]:
with open('radio_svoboda_scrape.py') as f:
    text = f.read()

In [22]:
print(text)

# -*- coding: utf-8 -*-
"""
Created on Thu Feb 21 01:14:12 2019

@author: Patrick Stoyer
"""

from urllib import request #to get html source code from websites
from bs4 import BeautifulSoup #to process source code
import pandas as pd
import bs4


sites_list = ['https://www.svoboda.org/z/16685?p={}'.format(x) for x in range(0,101)]
def get_article_links(page):
    articles_list = []
    response = request.urlopen(page)#request.urlopen returns html in bytes
    source = response.read().decode('utf-8') #decode bytes and append to source code list
    soup = BeautifulSoup(source, 'html.parser')
    for parent in soup.descendants: #for loop through all the soup's descendant
            if (parent.name =='li') and(parent.has_attr('class'))and (['col-xs-12', 'col-sm-12', 'col-md-12', 'col-lg-12'] == parent.get('class')):
                temp_stuff = ['','','','','',''] #this temporary list will hold various information to be stored
                for child in parent.descendants: #if the name

CODE
----
Here's an example of the code I used for scraping, this one was for radio svoboda...(ignore any of my comments that don't make sense, I copied a large portion of this from early code I used to scrape Amnesty International) I scrape each rubric page and collect each article's link, title, and date from their source... then I create a DataFrame from that list... I fix the titles' names so that they do not contain banned-from-filename characters... then I pass this to a for loop which writes the text to a file. originally i was trying to just put the text in the dataframe, but such a dataframe would take up an immense amount of memory. I also could have used a ufunc to write the files, but I wanted to keep track in the console of which article I was on and a forloop seemed like the most straightforward way to do this. Now let's take a closer look at the corpus.

In [25]:
bbc_corp = nltk.corpus.PlaintextCorpusReader('bbc_articles','.*\.txt')
kommersant_corp = nltk.corpus.PlaintextCorpusReader('kommersant_articles','.*\.txt')
radio_svoboda_corp = nltk.corpus.PlaintextCorpusReader('radio_svoboda_articles','.*\.txt')
reuters_corp = nltk.corpus.PlaintextCorpusReader('reuters_articles','.*\.txt')
tass_corp = nltk.corpus.PlaintextCorpusReader('tass_articles','.*\.txt')

In [26]:
len(bbc_corp.words())
len(kommersant_corp.words())
len(radio_svoboda_corp.words())
len(reuters_corp.words())
len(tass_corp.words())

336593

2055141

708533

91514

1067123

In [27]:
len(set(bbc_corp.words()))
len(set(kommersant_corp.words()))
len(set(radio_svoboda_corp.words()))
len(set(reuters_corp.words()))
len(set(tass_corp.words()))

41719

117152

57054

14776

79774

WORDCOUNT
---
By far the largest corpus is kommersant with over 2 million words (including punctuation I think) 117152 of which are unique. next largest are tass then radio_svoboda with 1 million and 700 thousand words respectively. Considering the number of articles, BBC is fairly large in comparison with the others, and the unique wordcount is also fairly sizeable. Then theres reuters, which is pretty pathetic.


In [38]:
text = bbc_corp.open(bbc_corp.fileids()[1]).read()
print(text+"\n\n")
len(text)
len(nltk.word_tokenize(text))
len(nltk.sent_tokenize(text))

Международная ассоциация легкоатлетических федераций (IAAF) разрешила 42 российским легкоатлетам участвовать в международных турнирах под нейтральным флагом, при этом Всероссийская федерация легкой атлетики (ВФЛА) по-прежнему дисквалифицирована, говорится в пресс-релизе IAAF.
Соответствующее решение приняла комиссия по допинг-контролю ассоциации, пришедшая к выводу, что 42 российских спортсмена выполнили все установленные требования и могут участвовать в соревнованиях 2019 года.
Рассмотрение заявок от российских спортсменов, желающих выступать под нейтральным флагом, началось в декабре 2018 года. В итоге были одобрены заявки 42 атлетов.
Среди тех, кому разрешили участвовать в турнирах этого года - двукратная чемпионка мира, прыгунья в высоту Мария Ласицкене, чемпион мира барьерист Сергей Шубенков, толкатель ядра Максим Афонин, бегун Ринас Ахмадеев, прыгунья в высоту Мария Кочанова.
В 2018 году 73 российских атлета получили разрешение выступать на международных турнирах в нейтральном ст

2625

397

20

EXAMPLE SCRAPE
----
For a second I thought this was in english but it turns out my browser was automatically translating it lol. As you can see, the article looks pretty good, I can't see any issues. This one is pretty short for the bbc corpus at about 5 kb...for comparison, here's the largest bbc file (only kommersant had any articles longer). 

In [41]:
text = bbc_corp.open('Я_идеальный_человек-Россия_рэпер_Фейс_от_Бургера_до_Мандельштама_6_сентября_2018.txt').read()
print(text+"\n\n")
len(text)
len(nltk.word_tokenize(text))
len(nltk.sent_tokenize(text))

18+
Иван Дремин, он же рэпер Фейс, в школе торговал гашишем, прославился треками, состоящими из мата, еще год назад тратил гонорары на одежду от Gucci. Почему теперь кумир российских школьников пишет песни о коррупции и пошел на оппозиционный "Марш матерей"?
"Я новый русский Тупак"
Июльским днем у дверей небольшой студии звукозаписи на Николоямской улице стоит телохранитель. Внутри, в полуподвальном помещении, 21-летний Иван Дремин, известный подросткам по всей России как рэпер Фейс, записывает последнюю песню для своего нового альбома. Голый по пояс, черные волосы в беспорядке, на пюпитре вместо бумаг - айфон. Дремин уже шестой раз перепевает рефрен заглавного трека.
- Я ворованный воздух! Я ворованный…  черт, заново… Я ворованный воздух! Вы просто мразь, а я ворованный воздух… Заново… Еще раз, теперь дабл...
Эти строчки - перефразированная цитата из манифеста Осипа Мандельштама "Четвертая проза" о цензуре: "Все произведения мировой литературы я делю на разрешенные и написанные без ра

42539

8395

509

This one is 509 sentences and over 8400 words... thats pretty big.
BBC was probably the most well formatted one since I did it last. 
Some of the other sites have issues, such as kommersant. When scraping that
one I forgot to write newline characters after each line so it's ugly.
Also, some have a bunch of the source code in them for some reason, that I 
need to find some way to fix:

In [43]:
kommersant_corp.open('Чечено-ингушскую_границу_изучили_со_всех_сторон_Конституционный_суд_рассмотрел_запрос_главы_Ингушетии_2018-11-27.txt').read()


'Конституционный суд (КС) РФ провел публичные слушания по запросу главы Ингушетии Юнус-Бека Евкурова о проверке соответствия Конституции РФ ратифицированного парламентом республики и аннулированного Конституционным судом Ингушетии соглашения с Чечней об установлении административной границы между двумя регионами. Руководителей Чечни и Ингушетии полностью поддержали представители федеральных органов власти, а ингушских оппозиционеров нейтрализовали их земляки из провластной общественности, включая имама, а также старейшины из Чечни.Дело о границе между Чечней и Ингушетией собрало в КС беспрецедентное число выступающих. Юнус-Бек Евкуров обратился в КС не «о признании неконституционности», а о подтверждении конституционности оспоренных актов «вопреки одному из высших органов Ингушетии», подчеркнул в начале заседания Валерий Зорькин. Напомним, КС Ингушетии 30 сентября признал нарушающими конституцию республики закон о ратификации соглашения, которым впервые была установлена граница с Чечне

(pretty ugly but probably not a major issue for computational tasks. The next one is much worse.)

In [44]:
kommersant_corp.open('Сергей_Собянин_оставил_позади_и_конкурентов,_и_самого_себя_образца_2013_года_Выборы_мэра_Москвы_не_2018-09-10.txt').read()



'На выборах мэра Москвы ожидаемо победил действующий градоначальник Сергей Собянин. По сравнению с 2013 годом его победа оказалась более убедительной: если тогда во второй тур едва не вышел кандидат РПР—ПАРНАС Алексей Навальный, то в этот раз мэр заручился поддержкой 70% проголосовавших. Явка составила 30,89% при 32,03% в 2013 году. По сравнению с прошлыми выборами вдвое увеличился процент недействительных бюллетеней — возможно, это связано с ограниченностью кандидатского предложения для избирателей, настроенных оппозиционно по отношению к господину Собянину.Мэр Москвы намерен «пахать» больше в последний срокПо результатам обработки 99% протоколов победу на выборах мэра Москвы одержал Сергей Собянин. Он получил 70,14% голосов (1,5\xa0млн из 7,2\xa0млн), следует из данных Мосгоризбиркома (МГИК). Явка избирателей на выборах в нынешнем году оказалась меньше, чем в 2013 году,— 30,89% (была 32,03%). В 2013 году, когда в выборах участвовал оппозиционер Алексей Навальный (набрал 27,24%, 632\x

I need to find some way to fix this, either by automatically removing it somehow, or by re-scraping these articles with a better function